RFC: Flash area changes

de-nordic opened the pull request:
This is proposal to flash_area changes that remove no longer need fields and replace default_flash_map with list of linkable flash_area objects.

It proposes replacement of direct access to flash_area fields, from non-zephyr internals, via macros (here, but these can also be functions) to hide implementation details.
The flash_area_open function has been changed to detect whether it has been given flash id or flash_area pointer and will search for appropriate area by id in first case and directly return the given pointer in second case.
The flash_area_open will not be used internally as flash area objects will be available by FLASH_AREA_BY_LABEL macro (the change has only be done partially)

The current changes have been tested with basic mcuboot builds and seem to work correctly.

- [ ] Documentation
- [ ] More mcuboot testing
- [x] Tests of mcumgr

The RFC has been provided in response to https://github.com/zephyrproject-rtos/zephyr/issues/34227 and lack of acceptance for https://github.com/zephyrproject-rtos/zephyr/pull/33346.

The intention is to cut down the API to what is really needed and avoid introduction of new API that would require compatibility layer, for example for mcuboot.

There is alternative solution provided here https://github.com/zephyrproject-rtos/zephyr/pull/34494 by @Laczen .
Alternative solution in form of flash partition devices, that use the same API as flash devices: https://github.com/zephyrproject-rtos/zephyr/pull/36487

de-nordic committed:
flash_area: Add linker section for flash_map

The commit adds linker ROM section for storage of flash map.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
storage/flash_area: Great overhaul of flash map API

The commit contains following changes to flash map API:
 - flash_area_open and flash_area_close have been marked as deprecated
   and no longer should be used within Zephyr code;
 - the fa_id, fa_dev_name, fa_device_id have removed from flash_area
   structure;
 - flash device, that the flash area is on, is now linked to the area
   at compile time;
 - functions that are conditionally compiled in with
   CONFIG_FLASH_PAGE_LAYOUT and CONFIG_FLASH_AREA_CHECK_INTEGRITY
   have been moved to dedicated units.
 - adds FLASH_AREA(label) macro that has to be invoked to obtain DTS
   defined area, to be directly used with read/write/erase functions;
   this replaces usage of FLASH_AREA_ID(label) with deprecated
   flash_area_open.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
storage/flash_map: Remove direct access to fa_dev_name

The commit modifies code that has been trying to directly access
fa_dev_name, of flash_area, to use fa_dev->name instead.
Any use of flash_area_get_device has been replaced with direct access
to flash_area.fa_dev.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
west.yaml: Update mcuboot required by flash_area changes

The mcuboot requires update to work with changes to flash_area.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
fs/fcb: Update fcb_init to work with new flash_area access

The commit alters the fcb_init function to accept -1 as area identifier
that indicates that the fcb.fap, provided to the function, already
contains flash_area object pointer that should be used to access
flash area.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/fs/fcb: Switch fcb_init calls to use flash_area object in fcb

The commit changes flash_init calls to use the flash_area object
pointers given via fcb.fap instead of trying to obtain the pointer
by flash area ID.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
settings: Switch fcb_init calls to use flash_area object in fcb

The commit changes flash_init calls to use the flash_area object
pointers given via fcb.fap instead of trying to obtain the pointer
by flash area ID.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nvlsianpu reviewed (commented):
Regard mcuboot api we should wait for the mcuboot PR review (expect no problem with that) but I think getter function should land in dedicated for MCUboot header.

nvlsianpu commented on include/storage/flash_map.h:
```diff
@@ -51,22 +54,37 @@ extern "C" {
+
+inline uint8_t flash_area_get_device_id(const struct flash_area *fa)
+{
+	return SOC_FLASH_0_ID;
```
This is used only on ARM cores, so is ok.

de-nordic committed:
usb/dfu: Switch from flash area IDs to directly access

The commit switches the USB DFU from using flash area IDs to obtain
access to flash_areas to directly accessing flash_area objects.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
dfu: mcuboot: Change to new flash_area access

The commit changes the mcuboot functions that have operated
on flash area ID to dump flash_area_open/flash_area close
where applicatble and use flash_area pointer instead of
flash area ID.
The commit provides also changes to dependent code in subsystems:
 can, dfu and mgmt.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
storage/flash_map: Add flash_area_erase_whole(...) function

The commit adds function that performs erase of whole flash_area
of given flash_area object.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
dts: gen_defines: Generate PARTITION_ID only for "okay" nodes

The commit modifies how the PARTITION_ID is generated, to provide
an ID only for partition nodes that are "okay".
The change will cause code trying to obtain ID of non-"okay"ed
partition to fail compilation, which allows compile time detection
of addressing non-existing partitions.
This also allows to produce, at compile time, arrays of objects
that can be indexed by the ID of partition, while not including
disabled nodes.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nvlsianpu commented on include/storage/flash_map.h:
```diff
@@ -262,15 +267,66 @@ uint8_t flash_area_erased_val(const struct flash_area *fa);
+ *
+ * Usage: DEFINE_FLASH_AREA(some) = { ... };
+ */
+#define DEFINE_FLASH_AREA(name)									\
```
to many whitespaces

nvlsianpu commented on subsys/storage/flash_map/flash_map.c:
```diff
@@ -34,41 +34,40 @@ struct layout_data {
+	if (area >= __flash_map_list_start && area < __flash_map_list_end) {
+		/* For id=>area to be valid flash_area pointer it has to be aligned as such */
+		if ((const uintptr_t)area & (const uintptr_t)__alignof(struct flash_area)) {
+			return -EFAULT;
```
return -EINVAL?

de-nordic commented on include/storage/flash_map.h:
```diff
@@ -262,15 +267,65 @@ uint8_t flash_area_erased_val(const struct flash_area *fa);
+ * Obtain pointer to flash_area objet directly from DTS label of fixed partition.
+ *
+ * @param[in] label Fixed partition label to use; this C identifier type label, which menas that
+ *	      instead of for example image-0, image_0 sh uld be provided.
```
> 
> 
> typo: sh uld -> should

Heh, must have had a short-circuit in brain to use space instead of o.

nvlsianpu reviewed (commented):
Patch looks fine to me. I'm not done with the review although have most of it (flash_area rework details) behind.
[EDIT]
I finished the review.

nvlsianpu commented on include/dfu/flash_img.h:
```diff
@@ -36,13 +36,26 @@ struct flash_img_check {
  *
  * @return  0 on success, negative errno code on fail
  */
 int flash_img_init_id(struct flash_img_context *ctx, uint8_t area_id);
```
need to design API using  `__deprecated`
This remark is also valid to any else deprecated API function.

de-nordic replied:
Actually I have left the comment but I am not sure about deprecation. Both functions can coexist, of course I would like people to use the `flash_img_init_area` rather than the `flash_img_init_id` but it does not seem to hurt to have both. Should we deprecate it @nvlsianpu ?

nvlsianpu replied:
You sand that the call is deprecated above - that why I requested the change.
Seams deprecation is not really your intention, let's change the comment to reflect that:
`This API call is kept for backward compatibility. flash_img_init_area() is recommended to be used instead.`

nvlsianpu commented on include/dfu/mcuboot.h:
```diff
@@ -73,6 +73,8 @@ extern "C" {
 						  BOOT_MAGIC_SZ -\
 						  BOOT_MAX_ALIGN * 2)
+
+struct flash_area;
```
Can you add comment in the code on that?

de-nordic commented on subsys/storage/flash_map/flash_map.c:
```diff
@@ -141,12 +140,11 @@ flash_page_cb cb, struct layout_data *cb_data)
 
-	flash_dev = device_get_binding(fa->fa_dev_name);
-	if (flash_dev == NULL) {
+	CHECKIF(fa->fa_dev == NULL) {
```
Yes. I might be wrong here: the idea is that since the flash_area object is generated at compile time, the `fa_dev` is expected to be attached or compilation error to occur. So there will always be some `fa_dev` , it might not be the right one if DST or custom flash map is not OK, but it will be there.

mbolivar-nordic commented on include/linker/common-rom.ld:
```diff
@@ -131,6 +131,15 @@
+#if defined(CONFIG_FLASH_MAP)
+	SECTION_DATA_PROLOGUE(_flash_map,,SUBALIGN(4))
+	{
+		__flash_map_list_start = .;
```
The commit log here says what is being done, but not why, so I'm not sure what this is for. The pull request description doesn't mention the linker either.

What is the goal? Why do we need a separate linker section? It's not clear why this is the first patch in the series.

de-nordic replied:
The separate linker section is used for ordering the areas so that code, outside of zephyr, that still uses id could be picking them by id (index). This also required change in gen_defines.py because only ID for "okay" partitions are now expected, so that they would be continuous.

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -354,9 +355,11 @@ def write_special_props(node):
-        macro = f"{node.z_path_id}_PARTITION_ID"
-        out_dt_define(macro, flash_area_num)
-        flash_area_num += 1
+        if node.status == "okay":
```
The reason for this change is also not explained in the commit log. It says what is happening, but not why.

What is the goal?

I'm confused because I spot-checked various nodes in the tree and they all look like this:

	partitions {
		compatible = "fixed-partitions";
		boot_partition: partition@0 {
			label = "mcuboot";
			reg = <0x000000000 0x0000C000>;
		};
                /* ... */
	};

Since boot_partition does not have an explicit status, it has status "okay", because that is the behavior when status is missing.

I couldn't find any nodes which have status "disabled" in this PR, so it doesn't seem to have any effect.

What am I missing?

de-nordic replied:
It is done so that partitions that get disabled do not get IDs as they will not be included into map.
The macros that include partitions into map add only these that have "okay" status, which means that when storing partitions in array, you may not use ID to index them because the ID is not related status.

mbolivar-nordic replied:
> It is done so that partitions that get disabled do not get IDs as they will not be included into map.

But no partitions are disabled, so I don't get the point.

mbolivar-nordic commented on include/storage/flash_map.h:
```diff
@@ -113,14 +115,17 @@ int flash_area_check_int_sha256(const struct flash_area *fa,
  * -ENOENT if @p ID is unknown.
  */
-int flash_area_open(uint8_t id, const struct flash_area **fa);
+int flash_area_open(const uintptr_t id, const struct flash_area **fa);
```
Changing flash_area_open() to accept a pointer feels a little weird and potentially a bit fragile.

It seems like we could add a separate function, like this:

```c
   /* maybe a name could be found than "..._ptr" */
   int flash_area_open_ptr(struct flash_area *area);
```

This would return 0 if `area` was successfully opened, and `-errno` otherwise.

Why do we have to make this change? The commit log says what is happening, but doesn't explain why.

Same question for `flash_area_get_sectors()`.

I'm guessing this is somehow for compatibility with mcuboot, but I thought one goal of this PR was to add some macros in mcuboot which wrap the Zephyr APIs with what MCUboot needs, so I don't get why this is needed in Zephyr; why is it?

mbolivar-nordic replied:
Another related question is why we are keeping around flash_area_open(), since we don't actually do anything on close(), instead of just replacing it with a  `flash_area_from_id` function like I ask about below. We could keep flash_area_open() around for compatibility, but deprecate it.

mbolivar-nordic commented on include/storage/flash_map.h:
```diff
@@ -262,15 +267,63 @@ uint8_t flash_area_erased_val(const struct flash_area *fa);
+ * @return const struct flash_area * to flash_area object or NULL.
+ */
+#define FLASH_AREA(label)								\
+	(&_CONCAT(_flash_map_area_,							\
```
Does this really return NULL on missing partitions? It looks like it would cause a linker error instead.

de-nordic replied:
> 
> 
> Does this really return NULL on missing partitions? It looks like it would cause a linker error instead.

Sorry, the comment was from previous version where it did, but it is better to let it fail at compilation/linking.

mbolivar-nordic commented on include/storage/flash_map.h:
```diff
@@ -262,15 +267,63 @@ uint8_t flash_area_erased_val(const struct flash_area *fa);
+
+
+#define DT_FLASH_AREA_FOREACH_OKAY_FIXED_PARTITION(fn)				\
+		    UTIL_CAT(DT_FOREACH_OKAY_INST_, fixed_partitions)(fn)
```
This is an abstraction violation of devicetree.h implementation details -- we shouldn't be using `DT_FOREACH_OKAY_INST_fixed_partitions` like this outside of devicetree.h or its sub-headers.

I see these uses:

 - in flash_map_default.c
- later on in flash_map.h, to declare partitions generated from DTS

The flash_map_default.c case already has defined `DT_DRV_COMPAT` to `fixed_partitions`, so it can simply use DT_INST_FOREACH_STATUS_OKAY without needing this new macro.

If this is really needed here (why?), then I we should instead add a new DT API macro, like this:

```
#define DT_FOREACH_STATUS_OKAY(compat, fn)			\
	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(compat),		\
		    (UTIL_CAT(DT_FOREACH_OKAY_INST_,		\
			      compat)(fn)),			\
		    ())
```

And then redefine

```
#define DT_INST_FOREACH_STATUS_OKAY(fn)				\
	DT_FOREACH_STATUS_OKAY(DT_DRV_COMPAT, fn)
```

to avoid the abstraction violation.

de-nordic replied:
I wanted to avoid adding `DT_DRV_COMPAT ` to header because a source that includes the `flash_map.h` maybe defining the name before or after inclusion of flash_map which may give error during compilation and if I undef it here, to define my own then the rest of a compilation unit will have either no definition when needed or incorrect one.

Maybe the `DT_INST_FOREACH_STATUS_OKAY` is required.

mbolivar-nordic replied:
> Maybe the `DT_INST_FOREACH_STATUS_OKAY` is required.

You meant `DT_FOREACH_STATUS_OKAY`, right?

mbolivar-nordic commented on subsys/storage/flash_map/flash_map.c:
```diff
@@ -24,7 +24,7 @@
 #if defined(CONFIG_FLASH_PAGE_LAYOUT)
 struct layout_data {
-	uint32_t area_idx;
+	const struct flash_area *area_idx;
```
Should be just `area` now, not `area_idx`, right?

mbolivar-nordic commented on subsys/storage/flash_map/flash_map.c:
```diff
@@ -176,11 +172,18 @@ static bool get_sectors_cb(const struct flash_pages_info *info, void *datav)
 	struct layout_data data;
+	const struct flash_area *fa;
+
+	int rc = flash_area_open(id, &fa);
```
Aren't you technically supposed to close these areas when you're done with them?

I know that the open/close thing is yet another mynewt legacy and zephyr has never done anything on close, but when I was more involved here, the idea was that close might do something sometime, so it had to be called each time.

Otherwise you might as well just have something like this within zephyr:

```
/**
 * @param fa_id flash area ID
 * @return a pointer to the flash_area, or NULL if it does not exist
 */
struct flash_area *flash_area_from_id(int fa_id);
```

mbolivar-nordic reviewed (commented):
Hey @de-nordic , please accept my apologies for the delayed review.

I'm trying to get my head around the solution today but I admit I'm getting tripped up in the first few patches. Rather than delay a response even longer I'm going to offer some questions on the first few patches. Maybe we can set up some time to discuss this in a telecon if I can't get my head around the rest tomorrow.

Thanks for taking this on!

de-nordic commented:
> 
> 
> Hey @de-nordic , please accept my apologies for the delayed review.
> 

No problem  I'd rather wait with it than have it rushed into  2.6.

de-nordic committed:
fs/littlefs: Switch from flash area IDs to flash area pointers

The commit switches LittleFS and code that uses the file system,
inside Zephyr, to use flash area object pointers instead of
using flash area IDs.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
debug/coredump: Replace flash area ID with pointers

References to flash area by ID has been replaced with direct access
to flash area objects by pointers.
Calls to flash_area_open and flash_area_close have been removed.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
settings: Remove flash_area_open

The usage of flash_area_open has been replaced with direct acquisition
of flash area object via FLASH_AREA macro.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/nvs: Remove flash_area_open

The flash_area_open has been replaced with direct acquisition of
flash area objects with FLASH_AREA macro.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/inte_crb: Remove flash_area_open and flash_area_close

The flash_area_open has been replaced by direct acquisition
of flash area object pointer with FLASH_AREA macro.
flash_area_close is no longer needed.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

mbolivar-nordic commented on include/storage/flash_map.h:
```diff
@@ -58,6 +59,38 @@ struct flash_area {
+ * The getter functions cover fields of flash_area from mcuboot not the internal implementation
+ * of Zephyr.
+ */
+static inline off_t flash_area_get_off(const struct flash_area *fa)
```
Do we really need any of these inside the zephyr repository itself?

All the other MCUboot ports except mynewt seem to put their flash area abstractions inside MCUboot:

https://github.com/mcu-tools/mcuboot/blob/main/boot/cypress/cy_flash_pal/cy_flash_map.c
https://github.com/mcu-tools/mcuboot/blob/main/boot/mbed/src/flash_map_backend.cpp

Mynewt is of course the exception because these APIs were already defined using the mynewt API.

Following the other ports by putting the "glue" between the zephyr native APIs and MCUboot's "mynewt style" view of the world seems better than putting the glue here in the zephyr API header.

Zephyr apps don't seem to need these definitions; they can access the structure directly if they are using the zephyr flash map API.

Am I missing something?

de-nordic replied:
No, you are not missing anything. I will move the functions.

mbolivar-nordic reviewed (commented):
I had a chance to look over your responses and read the rest of the PR as it relates to the flash_map.h API. I didn't look at the commits touching the in-tree users though.

Overall I think it's great that `struct flash_area` can  be reduced to a simple device + offset + size; that looks perfect.

I have some questions and concerns about some of the details that I've left comments about.

Basically these break down into:

- do we need the extra linker magic? a `struct flash_area *all_the_areas[DT_NUM_INST_STATUS_OKAY(fixed_partition)]` could potentially be created instead using DT_INST_FOREACH_STATUS_OKAY that just points to every area, and we could use the indexes in there instead. That's more wasteful of flash, but removes implementation complexity. Might not be worth it.
- still not sure about some of the APIs that remain; things like flash_area_open / flash_area_close still seem mynewt or mcuboot specific and maybe aren't really needed in zephyr at all anymore. Or maybe they could be deprecated over time and replaced with something that just returns the flash area. That style of API would not need any open/close semantics that we don't use at all in Zephyr (and which MCUboot doesn't really need either, but that's their problem)
- not sure about the new APIs that look mcuboot-specific either, when all the other MCUboot ports keep their glue layers in the mcuboot repo

de-nordic committed:
storage/flash_map: Add flash_pages_info support

The commit adds support for flash_pages_info directly from flash_area,
in similar way they work with flash drivers.
Following functions are reproducing flash driver API counterparts
behaviour, but within flash area objects:
 flash_area_get_page_info_by_idx
 flash_area_get_page_info_by_offs
 flash_area_get_page_count

In addition to above the following functions, that have no counterparts
within flash API, have been provided:
 flash_area_get_next_page_info

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
settings: Use flash_area_get_page_info

The commit replaces usage of flash_area_get_sectors with
flash_area_get_page_info.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
dfu/flash_img: Move from flash area ID to direct flash area access

The commit alters the flash_img API to abondon usage for flash area ID.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

yashi commented:
This proposal isn't really related to CAN bus.  Removed `area:CAN` label.

de-nordic commented:
This approach will probably not be followed, as it is desired to phase out Flash Map API from Zephyr.
